# Taller — Limpieza y Análisis de Datos con Pandas y NumPy

# Objetivo

Aplicar técnicas de limpieza, transformación y análisis de datos utilizando las
librerías Pandas y NumPy a partir de una base de datos con inconsistencias similares
a escenarios reales.

**Archivos entregados **

ventas_sucias_5000.csv

**Instrucciones generales **
• Trabajar en un notebook de Jupyter (.ipynb) o script .py  
• El código debe estar comentado y organizado  
• Cada respuesta debe incluir:  
• Código  
• Explicación breve del resultado obtenido  
• Justificar decisiones de limpieza cuando sea necesario  



**Parte 1 — Exploración de datos (Pandas) **

**Actividades **

**Cargar la base de datos  
Mostrar**

• Primeras filas  
• Información general del dataset  
• Resumen estadístico  


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import files
uloaded =files.upload()

Saving ventas_sucias_5000.csv to ventas_sucias_5000.csv


In [ ]:
#Mostrar primeras filas Dataframe
df= pd.read_csv(('/content/ventas_sucias_5000.csv'))
df.head()

,cliente,producto,precio,cantidad,pais,metodo_pago,fecha
0,Maria,Monitor,1326.0,NaN,peru,Efectivo,2024-01-01 00:00:00
1,Luisa,Laptop,55.0,2,chile,Tarjeta,2024-01-01 01:00:00
2,Carlos,Monitor,1203.0,9,Colombia,Efectivo,2024-01-01 02:00:00
3,Luisa,Monitor,1304.0,3,Perú,TRANSFERENCIA,2024-01-01 03:00:00
4,Luisa,Monitor,426.0,6,chile,Tarjeta,2024-01-01 04:00:00


In [ ]:
#Informacion General del dataset
df.shape

(5000, 7)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   cliente      5000 non-null   object 
 1   producto     5000 non-null   object 
 2   precio       4950 non-null   float64
 3   cantidad     4950 non-null   object 
 4   pais         5000 non-null   object 
 5   metodo_pago  5000 non-null   object 
 6   fecha        5000 non-null   object 
dtypes: float64(1), object(6)
memory usage: 273.6+ KB


In [ ]:
df.describe(include="all")

,cliente,producto,precio,cantidad,pais,metodo_pago,fecha
count,5000,5000,4950.000000,4950,5000,5000,5000
unique,6,5,NaN,10,7,4,4825
top,Ana,Mouse,NaN,2,chile,TRANSFERENCIA,2024-01-19 00:00:00
freq,870,1069,NaN,573,733,1276,5
mean,NaN,NaN,5053.823434,NaN,NaN,NaN,NaN
std,NaN,NaN,63379.982009,NaN,NaN,NaN,NaN
min,NaN,NaN,10.000000,NaN,NaN,NaN,NaN
25%,NaN,NaN,531.000000,NaN,NaN,NaN,NaN
50%,NaN,NaN,1027.000000,NaN,NaN,NaN,NaN
75%,NaN,NaN,1519.000000,NaN,NaN,NaN,NaN


In [ ]:
#Tipos de datos
df.dtypes

,0
cliente,object
producto,object
precio,float64
cantidad,object
pais,object
metodo_pago,object
fecha,object


In [ ]:
df.isnull().sum()

,0
cliente,0
producto,0
precio,50
cantidad,50
pais,0
metodo_pago,0
fecha,0


In [ ]:
#duplicados
df['pais'].unique()

array(['peru', 'chile', 'Colombia', 'Perú', 'COL', 'Chile', 'col'],
      dtype=object)

In [ ]:
df['producto'].unique()

array(['Monitor', 'Laptop', 'Celular', 'Mouse', 'Teclado'], dtype=object)

In [ ]:
df['metodo_pago'].unique()

array(['Efectivo', 'Tarjeta', 'TRANSFERENCIA', 'transferencia'],
      dtype=object)

# **Resumen estadistico**

**Debes responder:**
1. ¿Cuántas filas y columnas tiene la base de datos?

**Respuesta:** La base de datos tiene 5000 filas y 7 columnas*

2. ¿Qué tipos de datos identificas?
**Respuesta:** Numérico (float64): Únicamente la columna precio.

Texto / Objeto (object): Las columnas cliente, producto, cantidad, pais, metodo_pago y fecha.

3. ¿Encuentras columnas con problemas o inconsistencias?
**Respuesta:**  

*   Al revisar df. info() el dataset tiene 5,000 registros pero en la columna precio y cantidad solo tiene 4,950 valores

*   En la columna cantidad esta clasificada como objet deberia estar como enteros

*   Hay duplicidad en las columnas pais, metodos de pago, producto





Limpieza de Datos, duplicado, nulos

**Realiza las siguientes tareas:**

• Corregir tipos de datos incorrectos
• Identificar y manejar valores nulos
• Estandarizar columnas de texto:
• País
• Método de pago
• Revisar posibles valores atípicos (outliers)
• Corregir formatos de fecha si es necesario

In [ ]:
df_limpio =df.copy()

In [ ]:
# convertir vacios y todo en minuscula
df_limpio["producto"]= df_limpio["producto"].str.lower().str.strip()
df_limpio["pais"]= df_limpio["pais"].str.lower().str.strip()
df_limpio["metodo_pago"]= df_limpio["metodo_pago"].str.lower().str.strip()

In [ ]:
#estandarizando
df_limpio["pais"]=df_limpio["pais"].replace({
 "perú":"peru",
"Col":"colombia",
"col":"colombia"
})

df_limpio["metodo_pago"]=df_limpio["metodo_pago"].replace({
"TRANSFERENCIA":"transferencia"

})


In [ ]:
#convertir a numero
df_limpio['cantidad']=pd.to_numeric(df_limpio["cantidad"],errors="coerce")
df_limpio['precio']=pd.to_numeric(df_limpio["precio"],errors="coerce")

#reemplazar celdas vacias con la mediana
df_limpio["precio"]=df_limpio["precio"].fillna(df_limpio["precio"].median())
df_limpio["cantidad"]=df_limpio["cantidad"].fillna(df_limpio["cantidad"].median())


In [ ]:
#pasar formato de fecha
df_limpio["fecha"]=pd.to_datetime(df_limpio["fecha"],errors="coerce")
df_limpio=df_limpio.dropna(subset=["fecha"])

In [ ]:
#valores atipicos (outliers)
df_limpio= df_limpio[df_limpio["precio"]<100000]

In [ ]:
#verificacion final
print("--- Diagnóstico de limpieza completo ---")
print(df_limpio.info())

--- Diagnóstico de limpieza completo ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   cliente      5000 non-null   object        
 1   producto     5000 non-null   object        
 2   precio       5000 non-null   float64       
 3   cantidad     5000 non-null   float64       
 4   pais         5000 non-null   object        
 5   metodo_pago  5000 non-null   object        
 6   fecha        5000 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(2), object(4)
memory usage: 273.6+ KB
None


In [ ]:
df_limpio.isnull().sum()

,0
cliente,0
producto,0
precio,0
cantidad,0
pais,0
metodo_pago,0
fecha,0


In [ ]:
df_limpio = df_limpio.reset_index(drop=True)

In [ ]:
df_limpio.head()

,cliente,producto,precio,cantidad,pais,metodo_pago,fecha
0,Maria,monitor,1326.0,5.0,peru,efectivo,2024-01-01 00:00:00
1,Luisa,laptop,55.0,2.0,chile,tarjeta,2024-01-01 01:00:00
2,Carlos,monitor,1203.0,9.0,colombia,efectivo,2024-01-01 02:00:00
3,Luisa,monitor,1304.0,3.0,peru,transferencia,2024-01-01 03:00:00
4,Luisa,monitor,426.0,6.0,chile,tarjeta,2024-01-01 04:00:00


**Debes responder:**
¿Qué problemas encontraste?
*Respuesta:* Valores faltantes y nulos, inconsistencia en precio y cantidad, tipos de datos incorrectos, valores atipicos

¿Cómo los solucionaste?

*Respuesta:* reemplazar celdas vacias, estandarizar datos, cambiar tipos de datos a los correctos

**¿Eliminaste registros? ¿Por qué?**
*Respuesta:* se elimino un registro de 999.999 debido a que distorcionaba los datos


**Parte 3 — Análisis con Pandas
Actividades**

• Crear una nueva columna llamada total: df['total'] = df['precio'] * df['cantidad']

Obtener:

• Total vendido

• Promedio de ventas

• Venta máxima

• Venta mínima

**Identificar:**

Los 5 productos con mayor valor vendido
El país con más ventas

In [ ]:
df_limpio['total'] = df_limpio['precio'] * df_limpio['cantidad']
print(f"Total vendido: {df_limpio['total'].sum():,.0f}")
promedio = df_limpio['total'].mean()
print(f"Promedio de ventas: {  promedio:.0f}")
maximo = df_limpio['total'].max()
print(f"Venta máxima:{maximo:.0f}")
minima=df_limpio['total'].min()
print(f"Venta mínima:{ minima:.0f}")

Total vendido: 25,221,047
Promedio de ventas: 5044
Venta máxima:17955
Venta mínima:10


In [ ]:
#agrupar productos
productos_top = df_limpio.groupby('producto')['total'].sum()
productos_top = productos_top.sort_values(ascending=False).head(5)
print("--- TOP 5 PRODUCTOS CON MÁS VENTAS ---")
for producto, valor in productos_top.items():
    print(f"Producto: {producto:<20} Total Vendido: ${valor:,.0f}")

--- TOP 5 PRODUCTOS CON MÁS VENTAS ---
Producto: mouse                Total Vendido: $5,425,150
Producto: laptop               Total Vendido: $5,146,842
Producto: monitor              Total Vendido: $4,967,368
Producto: celular              Total Vendido: $4,899,899
Producto: teclado              Total Vendido: $4,781,788


In [ ]:
#agrupar pais
ventas_por_pais = df_limpio.groupby('pais')['total'].sum()
pais_con_mayor_ventas = ventas_por_pais.idxmax()
monto_maximo = ventas_por_pais.max()
print(f"El país con más ventas es {pais_con_mayor_ventas:} con un monto total de ${monto_maximo:,.0f}")

El país con más ventas es colombia con un monto total de $10,918,498


**Parte 4 — Introducción a NumPy
Actividades**

• Convertir columnas numéricas a un array de NumPy: data = df[['precio',
'cantidad']].to_numpy()

• Separar columnas utilizando indexación: precios = data[:, 0]
cantidades = data[:, 1]

• Aplicar vectorización para calcular ventas totales: totales = precios *
cantidades

In [ ]:
ventas = df_limpio[['precio', 'cantidad']].to_numpy()
#estraer las columnas
precios = ventas[:, 0]
cantidades = ventas[:, 1]
#multiplicamos
total_ventas = precios * cantidades
print(f"Promedio ventas{ np.mean(total_ventas):.0f}")
print("Ventas mayores a 5000: ", len (total_ventas[total_ventas>5000]))
valor_ventas_mayores = np.sum(total_ventas[total_ventas > 5000])
print(f"Valor total de ventas mayores a 5000: ${valor_ventas_mayores:,.2f}")

Promedio ventas5044
Ventas mayores a 5000:  2070
Valor total de ventas mayores a 5000: $18,827,221.00


**Parte 5 — Análisis con NumPy
Utilizando funciones de NumPy, calcular:**

• Suma total de ventas

• Promedio de ventas

• Venta máxima

• Cantidad de ventas superiores a 1000  

In [ ]:
print(f"Suma total: ${np.sum(total_ventas):,.0f}")
print(f"Promedio de ventas: ${np.mean(total_ventas):,.0f}")
print(f"Venta maxima: ${np.max(total_ventas):,.0f}")
print(f"Cantidad de ventas superiores a 1000: ${np.sum(total_ventas[total_ventas>1000]):,.0f}")

Suma total: $25,221,047
Promedio de ventas: $5,044
Venta maxima: $17,955
Cantidad de ventas superiores a 1000: $24,831,829


**Debes responder:**

1. ¿Qué ventajas observas al usar NumPy?
*Respuesta:* Es mucho mas rapido al hacer los calculos y su eficiencia y el consumo de memoria es menor

2. ¿Qué significa “vectorización”?
*Respuesta:* es poder hacer operaciones en un solo conjunto y que sea mucho mas rapido

3. ¿Qué hace la expresión data[:, 0]?
*Respuesta:* es para seleccionar las una columna completa de una matriz

**Parte 6 — Interpretación de resultados
Analiza y responde:**

1. ¿Los resultados obtenidos tienen sentido?
*Respuesta* Si tienen sentido debido a que se realizo la limpieza de los datos y estos concuerdan con los resultados esperados.
2. ¿Detectaste valores sospechosos?
*Respuesta:* Si se detectaron valores muy grandes que no coincidian con la ventas
3. ¿El promedio representa
correctamente los datos?
*Respuesta:* cuando hay valores atipicos es la mejor opcion, para que no se me distorcione los datos.
4. ¿Qué decisiones tomarías si esta fuera información real de negocio?
*Respuesta:* Me permite comprar deacuerdo al producto que mas tiene rotacion, tener extrategias de los paises que menos vende, saber cual es porcentaje mayor que mas me genera rentabilidad